File: python/leetcode/310.py
class Solution:
    def findMinHeightTrees(self, n, edges):
        if n == 1:
            return [0]
        graph = [set() for _ in xrange(n)]
        for i, j in edges:
            graph[i].add(j)
            graph[j].add(i)
        leaves = [i for i in xrange(n) if len(graph[i]) == 1]
        while n > 2:  # remove all leaves
            n -= len(leaves)
            new_leaves = []
            for i in leaves:
                j = graph[i].pop()
                graph[j].remove(i)
                if len(graph[j]) == 1:
                    new_leaves.append(j)
            leaves = new_leaves
        return leaves

File: cpp/leetcode/310.cpp
class Solution {
public:
    vector<int> findMinHeightTrees(int n, vector<pair<int, int>> &edges) {
        if (n == 1) {
            return {0};
        }
        unordered_map<int, unordered_set<int>> graph(n);
        for (auto edge : edges) {
            graph[edge.first].insert(edge.second);
            graph[edge.second].insert(edge.first);
        }
        vector<int> leaves, newLeaves;
        for (auto it : graph) {
            if (it.second.size() == 1) {
                leaves.push_back(it.first);
            }
        }
        while (n > 2) {
            n -= leaves.size();
            newLeaves.clear();
            for (int i = 0; i < leaves.size(); i++) {
                int node = *graph[leaves[i]].begin();
                graph.erase(leaves[i]);
                graph[node].erase(leaves[i]);
                if (graph[node].size() == 1) {
                    newLeaves.push_back(node);
                }
            }
            swap(leaves, newLeaves);
        }
        return leaves;
    }
};